File: 2017/day22/day22.c
#include <stdio.h>

#define TARGET_X 770
#define TARGET_Y 7

void main()
{
  unsigned long	long cave[TARGET_X + 1][TARGET_Y + 1];
  unsigned long	long erosion[TARGET_X + 1][TARGET_Y + 1];
  int	depth = 10647;
  //  int	depth = 510;
  int	risk;
  int	x;
  int	y;

  risk = 0;
  x = 0;
  while (x <= TARGET_X)
    {
      y = 0;
      while (y <= TARGET_Y)
	{
	  if (x == 0 || y == 0)
	    cave[x][y] = x * 48271 + y * 16807;
	  else if (x == TARGET_X && y == TARGET_Y)
	    cave[x][y] = 0;
	  else
	    cave[x][y] = erosion[x - 1][y] * erosion[x][y - 1];
	  erosion[x][y] = (cave[x][y] + depth) % 20183;
	  risk += erosion[x][y] % 3;
	  ++y;
	}
      ++x;
    }
  printf("%d\n", risk);
}

File: 2017/day22/day22.py
def star1():
    depth = 10647
    target = [7, 770]
    cave = [[0] * (target[1] + 1)] * (target[0] + 1)
    erosion = [[0] * (target[1] + 1)] * (target[0] + 1)
    risk = 0
    for x in range(target[0] + 1):
        for y in range(target[1] + 1):
            if x == 0 or y == 0:
                cave[x][y] = 48271 * y + 16807 * x
            elif x == target[0] and y == target[1]:
                cave[x][y] = 0
            else:
                cave[x][y] = erosion[x - 1][y] * erosion[x][y - 1];
            erosion[x][y] = (cave[x][y] + depth) % 20183
            risk += erosion[x][y] % 3
    return risk

print(star1())
